lib/DateTime.pm

```diff
@@ -1828,7 +1828,13 @@ sub _compare_overload {
 
     # note: $_[1]->compare( $_[0] ) is an error when $_[1] is not a
     # DateTime (such as the INFINITY value)
-    return $_[2] ? -$_[0]->compare( $_[1] ) : $_[0]->compare( $_[1] );
+
+    if ( defined $_[0] && defined $_[1] ) {
+        return $_[2] ? -$_[0]->compare( $_[1] ) : $_[0]->compare( $_[1] );
+    }
+    else {
+        return $_[2] ? $_[0]->compare( $_[1] ) : $_[0]->compare( $_[1] );
+    }
 }
```

Comment:
This fix is more complicated than it needs to be. All you need to do is add this to `_compare_overload`:

    return undef unless defined $_[1];

The `_compare_ method that we ultimately call just returns `undef` when the second argument is undefined anyway.